src/inmanta/server/services/resourceservice.py

```diff
@@ -563,31 +612,46 @@ async def get_increment(
         :param version: The version of the model to consider.
         :param connection: connection to use towards the DB.
             When the connection is in a transaction, we will always invalidate the cache
+        :param run_ahead_lock: lock used to keep agents hanging while building up the latest version
         """
 
-        def _get_cache_entry() -> Optional[tuple[abc.Set[ResourceIdStr], abc.Set[ResourceIdStr]]]:
+        async def _get_cache_entry() -> (
+            Optional[tuple[int, abc.Set[ResourceIdStr], abc.Set[ResourceIdStr], abc.Mapping[str, abc.Set[ResourceIdStr]]]]
+        ):
             """
-            Returns a tuple (increment, negative_increment) if a cache entry exists for the given environment and version
+            Returns a tuple (increment, negative_increment, negative_increment_per_agent)
+            if a cache entry exists for the given environment and version
             or None if no such cache entry exists.
             """
             cache_entry = self._increment_cache.get(env.id, None)
             if cache_entry is None:
                 # No cache entry found
                 return None
-            (version_cache_entry, incr, neg_incr) = cache_entry
-            if version_cache_entry != version:
+            (version_cache_entry, incr, neg_incr, neg_incr_per_agent, run_ahead_lock) = cache_entry
+            if version_cache_entry > version:
+                # Cache is ahead,
+                if run_ahead_lock is not None:
+                    await run_ahead_lock.wait()
```

Comment:
Took me a while to realize this `run_ahead_lock` shadows the one passed in to the method. Perhaps we can rename it `cache_run_ahead_lock`?
Also added a comment.

```suggestion
            (version_cache_entry, incr, neg_incr, neg_incr_per_agent, cache_run_ahead_lock) = cache_entry
            if version_cache_entry > version:
                # Cache is ahead,
                if cache_run_ahead_lock is not None:
                    # wait for cache entry to be complete
                    await cache_run_ahead_lock.wait()
```